compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/kotlin/KSDefNonNullReferenceImpl.kt

```diff
@@ -0,0 +1,36 @@
+package com.google.devtools.ksp.symbol.impl.kotlin
+
+import com.google.devtools.ksp.KSObjectCache
+import com.google.devtools.ksp.findParentOfType
+import com.google.devtools.ksp.symbol.*
+import com.google.devtools.ksp.symbol.impl.toLocation
+import org.jetbrains.kotlin.psi.KtIntersectionType
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.psi.KtUserType
+
+class KSDefNonNullReferenceImpl private constructor(val ktIntersectionType: KtIntersectionType) :
+    KSDefNonNullReference {
+    companion object : KSObjectCache<KtIntersectionType, KSDefNonNullReferenceImpl>() {
+        fun getCached(ktIntersectionType: KtIntersectionType) = KSDefNonNullReferenceImpl
+            .cache.getOrPut(ktIntersectionType) { KSDefNonNullReferenceImpl(ktIntersectionType) }
+    }
+
+    override val enclosedType: KSClassifierReference by lazy {
+        (ktIntersectionType.getLeftTypeRef()?.typeElement as KtUserType).let { KSClassifierReferenceImpl.getCached(it) }
```

Comment:
This can throw cast exceptions which are implementation specific.  Please check the type and throw `IllegalStateException` like that in `KSTypeReferenceImpl.element`.